README.md
riscvOVPsim
===
A Complete, Fully Functional, Configurable RISC-V Simulator
===

Author: Imperas Software, Ltd., using OVP Open Standard APIs  
Date   : 22 Jul 2020 
Version: 20200722.0  
License: Model source included under Apache 2.0 open source license  
License: Simulator riscvOVPsim licensed under Open Virtual Platforms (OVP) Fixed Platform Kits license  
RISC-V Specifications currently supported:
- RISC-V Instruction Set Manual, Volume I: User-Level ISA (User Architecture Version 20190305-Base-Ratification)
- RISC-V Instruction Set Manual, Volume II: Privileged Architecture (Privileged Architecture Version 20190405-Priv-MSU-Ratification)
- RISC-V Instruction Set Manual, RISC-V "V" Vector Extension (with version configurable in the model using the 'vector_version' parameter. 'master' version conforms to specification changes up to 3 July 2020 and is regularly updated to track the evolving specification.)
- RISCV Extension B (Bit Manipulation)
  - Run command: riscvOVPsim.exe --override riscvOVPsim/cpu/add_Extensions=B --showoverrides
  - When B extension is active, parameter 'bitmanip_version' can be used to specify the required version.
  
About riscvOVPsim
---
The **riscvOVPsim** simulator implements the full and complete functionality of the RISC-V Foundation's public User and Privilege specifications.  

The simulator is command line configurable to enable/disable all current optional and processor specific options. 

The simulator is developed, licensed and maintained by [Imperas Software](http://www.imperas.com/riscv) and it is fully compliant to the OVP open standard APIs. 

As a member of the RISC-V Foundation community of software and hardware innovators collaboratively driving RISC-V adoption, Imperas has developed the riscvOVPsim simulator to assist RISC-V adopters to become compliant to the RISC-V specifications. The latest RISC-V compliance test suite and framework can be downloaded from https://www.github.com/riscv/riscv-compliance . 

riscvOVPsim includes an industrial quality model and simulator of RISC-V processors for use for compliance and test development. It has been developed for personal, academic, or commercial use, and the model is provided as open source under the Apache 2.0 license. The simulator is provided under the  Open Virtual Platforms (OVP) Fixed Platform Kits license that enables download and usage. riscvOVPsim and Imperas RISC-V support is actively maintained and enhanced. To ensure you make use of the current version of riscvOVPsim this initial release will expire. Please download the latest version.

Runtime configurable settings for all RISC-V specification options makes it very easy to compare run time results with any RTL implementations.

Full commercial features including variant selection, comprehensive trace, GDB/Eclipse graphical source code debug, semihosting, and RISC-V foundation signature dump facility for compliance testing.

More information: [riscvOVPsim user guide](doc/riscvOVPsim_User_Guide.pdf)  
Model Source: [source](source)  
Examples: [examples](examples)  

![](riscvOVPsim.jpg)  
riscvOVPsim is a fixed function simulation of one configurable processor model in a fixed platform. Full extendable platform simulations of reference designs booting FreeRTOS, Linux, SMP Linux etc. are available as open source and are available from [www.IMPERAS.com](http://www.imperas.com), [www.OVPworld.org](http://www.OVPworld.org).  


Debugging using GDB / Eclipse
---
The same fixed platform can be used to debug the application using GDB and Eclipse.

For GDB debug add the command line argument _-gdbconsole_, this will start the GDB debugger and connect to the virtual platform.

For Eclipse debug, using CDT, a configuration to attach to an external remote application for debug should be made. The port to attach to is opened on the running virtual platform by adding the command line argument _-port <port number>_ when starting the virtual platform. 

Please see the [riscvOVPsim user guide](doc/riscvOVPsim_User_Guide.pdf) for full details.

Using riscvOVPsim
---
To use the simulator, just download the files, go into one of the example directories, and execute the provided run scripts.
For example on Linux:  

> $ cd examples  
> $ cd fibonacci  
> $ RUN_RV32_fibonacci.sh  
> 
> CpuManagerFixedPlatform (64-Bit) 20180425.0 Open Virtual Platform simulator from [www.IMPERAS.com](http://www.imperas.com).  
> Copyright (c) 2005-2020 Imperas Software Ltd.  Contains Imperas Proprietary Information.  
> Licensed Software, All Rights Reserved.  
> Visit [www.IMPERAS.com](http://www.imperas.com) for multicore debug, verification and analysis solutions.  
>   
> CpuManagerFixedPlatform started: Tue Apr 24 19:12:06 2018  
>   
> Info (OR_OF) Target 'riscvOVPsim/cpu' has object file read from 'fibonacci.RISCV32-O0-g.elf'  
> Info (OR_PH) Program Headers:  
> Info (OR_PH) Type           Offset     VirtAddr   PhysAddr   FileSiz    MemSiz     Flags Align  
> Info (OR_PD) LOAD           0x00000000 0x00010000 0x00010000 0x00016998 0x00016998 R-E   1000  
> Info (OR_PD) LOAD           0x00017000 0x00027000 0x00027000 0x000009c0 0x00000a54 RW-   1000  
> starting fib(38)...  
> fib(0) = 0  
> fib(1) = 1  
> fib(2) = 1  
> fib(3) = 2  
> ...  
> fib(36) = 14930352  
> fib(37) = 24157817  
> finishing...  
> Info   
> Info ---------------------------------------------------  
> Info CPU 'riscvOVPsim/cpu' STATISTICS  
> Info   Type                  : riscv (RVB32I+MACSU)  
> Info   Nominal MIPS          : 100  
> Info   Final program counter : 0x100ac  
> Info   Simulated instructions: 4,400,537,204  
> Info   Simulated MIPS        : 1439.2  
> Info ---------------------------------------------------  
> Info   
> Info ---------------------------------------------------  
> Info SIMULATION TIME STATISTICS  
> Info   Simulated time        : 44.01 seconds  
> Info   User time             : 3.06 seconds  
> Info   System time           : 0.00 seconds  
> Info   Elapsed time          : 3.10 seconds  
> Info   Real time ratio       : 14.18x faster  
> Info ---------------------------------------------------  
>   
> CpuManagerFixedPlatform finished: Tue Apr 24 19:12:10 2018  
>   
> CpuManagerFixedPlatform (64-Bit) 20180425.0 Open Virtual Platform simulator from [www.IMPERAS.com](http://www.imperas.com).  
> Visit [www.IMPERAS.com](http://www.imperas.com) for multicore debug, verification and analysis solutions.  
  
Measuring Instruction Functional Coverage with the riscvOVPsim coverage engine
---
Instruction Functional Coverage as it relates to processor verification is a technology solution to measure what is being stimulated in the ISA in terms of which instructions, operands and values are driven into a processor.

The Imperas coverage technology is developed using the Imperas VAP intercept technology and is provided as part of riscvOVPsim.

Example coverage command:

> $ riscvOVPsim.exe --variant RVB32I --program eg.elf --cover basic --extensions RVI --reportfile impCov.log

The Imperas Instruction Functional Coverage works by monitoring every instruction as it retires and recording information about it.

At the end of simulation this data is summarized in the console and simulation log file.

> COVERAGE :: RVI :: threshold : 1 : instructions: seen 13/40 :  32.50%, coverage points hit: 262/2952 :   8.88%

Full details are provided in the report file.

You can get coverage reports for one run or merge results from many runs. You can select which instruction 
extension to report on, or even select just individual instructions to report on.

There is currently 'basic' and 'extended' coverage measuring.

Please read the [riscvOVPsim user guide](doc/riscvOVPsim_User_Guide.pdf) for full operational instructions.

If a signature comparison based verification methodology is adopted (as in the RISC-V Compliance suites) for comparison 
between device under test and reference, then functional coverage is only part of the story, 
as it is essential to measure the successful propagation of the results of the input instructions/values into the signature. 
Read more about this in the section below on Mutation Testing in the [riscvOVPsim user guide](doc/riscvOVPsim_User_Guide.pdf).

The coverage technology provided as part of riscvOVPsim is fixed functionality. The Imperas coverage technology is available as an extension library as source as a standard part of the Imperas commercial
 product offerings and this allows users to extend and modify functionality and coverage capability. Contact Imperas for more information.

Extending riscvOVPsim and building your own models and platforms
---
riscvOVPsim is a fixed function simulation of one configurable processor model in a fixed platform. 
Full extendable platform simulations of reference designs booting FreeRTOS, Linux, SMP Linux etc. 
are available as open source and are available from [www.IMPERAS.com](http://www.imperas.com), 
[www.OVPworld.org](http://www.OVPworld.org).  

About Open Virtual Platforms (OVP) and Imperas Software
---
**Open Virtual Platforms** was created in 2008 to provide an open standard set of APIs and methodology to develop virtual platforms and simulation technology. 
[www.OVPworld.org](http://www.OVPworld.org/riscv).

**Imperas Software Ltd.** is the leading independent commercial developer of virtual platforms and high-performance software simulation solutions for embedded processor and systems. Leading semiconductor and embedded software companies use Imperas simulators for their processor based simulation solutions.
[www.imperas.com](http://www.imperas.com/riscv).

![OVP Image ](http://www.imperas.com/sites/default/files/partner-logos/ovp_0.jpg)  
![Imperas Imperas](https://www.imperas.com/sites/default/files/imperas-web-logo_2.png)  
---


This is the riscvOVPsim/README.md  


ChangeLog.md
riscvOVPsim Change Log
===
Copyright (c) 2005-2020 Imperas Software Ltd., www.imperas.com

This CHANGELOG contains information for the riscvOVPsim fixed platform which includes information of the OVP Simulator and RISCV processor model

---

NOTE: X-commit messages below refer to git commits in the following
      Risc-V specification document repositories:
  I-commit: https://github.com/riscv/riscv-isa-manual
  V-commit: https://github.com/riscv/riscv-v-spec

- Vector Extension
  - V-commit b8cd98b: CSR vtype format changed to make vlmul bits contiguous.

Date 2020-July-21
Release 20200720.0
===

- First release of Imperas Instruction Functional Coverage engine to provide
  coverage of RISC-V Compliance Test Suites.
- mintstatus, sintstatus and uintstatus CSRs have been reassigned to addresses
  0x346, 0x146 and 0x046, respectively.
- The vector version master branch currently has these differences compared to
  the previous 0.9 version:
  - V-commit a679250: instruction vrgatherei16.vv added.

Date 2020-July-09
Release 20200708.0
===

- Reads of instret and cycle CSRs now exclude the current instruction from the
  reported count.
- The vector version master branch currently has these differences compared to
  the previous 0.9 version:
  - V-commit 7facdcc: SLEN=VLEN register layout is mandatory;
  - V-commit b454810: ELEN>VLEN is now supported for LMUL>1;
  - V-commit 20f673c: Whole register moves and load/stores now have element size
    hints;
  - V-commit d35b23f: Instructions vfrsqrte7.v and vfrece7.v added, with 
    candidate implementations (precise behavior is not yet defined).

Date 2020-June-30
Release 20200629.0
===

- Vector Extension
  - behavior of vslidedown.vx and vslidedown.vi with a slide of 0 when source
    and destination vector registers are the same has been corrected.

Date 2020-June-22
Release 20200619.0
===

- Vector Extension
  - sstatus.VS field alias has been implemented in its new position (from
    specification version 0.9).

Date 2020-June-17
Release 20200616.0
===

- Core-Local Interrupt Controller (CLIC)
  - address for xintthresh CSRs have been changed to 0xm47 (previously 0xm4A)
- Vector Extension
  - Checking of overlap of vector registers for vector indexed segment loads and
    stores has been corrected.

Date 2020-June-09
Release 20200608.0
===

- The Bit-Manipulation Extension is now implemented in the model, with version
  defined by parameter bitmanip_version.
- Some Vector Extension issues have been corrected:
  - Parameter order shown by disassembly of vector AMO instructions has been
    corrected. Model behavior is not affected by this change.
  - Encodings of integer extension instructions have been corrected for Vector
    Extension version 0.9.
- The vector version master branch currently has these differences compared to
  the previous 0.9 version:
  - V-commit 443ce5b: overlap constraints for different source/destination EEW
    changed.

Date 2020-May-27
Release 20200526.0
===

- Memory accesses that straddle PMP region boundaries are now disallowed for
  M-mode, even if those regions imply full M-mode access.

Date 2020-May-22
Release 20200521.0
===

- Memory accesses that straddle PMP region boundaries are now disallowed.
- EBREAK now sets tval to 0 if priv_version is set to master.
- Some Vector Extension issues have been corrected:
  - Encodings of VFUNARY0/VFUNARY1 instructions have been corrected for Vector
    Extension version 0.9.
  - Alignment of vector register groups when explicit EEW is being used has been
    corrected for Vector Extension version 0.9.

Date 2020-May-19
Release 20200518.0
===

- EBREAK and ECALL no longer count as retired instructions if priv_version
  is set to master.
- SC instruction behavior has been changed so that store address alignment is
  validated even if the reservation check fails (previously, alignment was
  validated only if the reservation check succeeded).
- Masking of the mcountinhibit CSR has been corrected - previously, bit 1 and
  bits 63...32 were writeable when they should have been 0.
- The nmi signal has been corrected to match the documented behavior, execution
  resumes at the nmi_address parameter value when the nmi signal goes high
- The optional Core-Local Interrupt Controller (CLIC) has been implemented
  (version 0.9-draft-20191208).
- Vector version 0.9 has been added, and is now used by default. Differences
  compared to the previous 0.8 version are as follows (with the associated
  specification V-commit identifiers):
  - V-commit bdb8b55: mstatus.VS and sstatus.VS fields have moved to bits 10:9;
  - V-commit b25b643: new CSR vcsr has been added and fields VXSAT and VXRM
    fields relocated there from CSR fcsr;
  - V-commit 951b64f: mirrors of fcsr fields have been removed from vcsr.
  - V-commit 1aceea2: vfslide1up.vf and vfslide1down.vf instructions added.
  - V-commit e256d65: vfcvt.rtz.xu.f.v, vfcvt.rtz.x.f.v, vfwcvt.rtz.xu.f.v,
    vfwcvt.rtz.x.f.v, vfncvt.rtz.xu.f.v and vfncvt.rtz.x.f.v instructions added;
  - V-commit 8a9fbce (and others): fractional LMUL support added, controlled by
    an extended vtype.vlmul CSR field;
  - V-commit f414f4d (and others): vector tail agnostic and vector mask agnostic
    fields added to the vtype CSR;
  - V-commit a526fb9 (and others): all vector load/store instructions replaced
    with new instructions that explicitly encode EEW of data or index;
  - V-commit ef531ea: whole register load and store operation encodings changed;
  - V-commit bdc85cd: vzext.vf2, vsext.vf2, vzext.vf4, vsext.vf4, vzext.vf8 and
     vsext.vf8 instructions added;
  - V-commit 9a77e12: MLEN is always 1.
- Some Vector Extension issues have been corrected:
  - Instructions vfmv.s.f and vfmv.f.s now require that SEW is a supported
    floating point size (pending vector specification clarification).


source/riscvUtils.c
/*
 * Copyright (c) 2005-2020 Imperas Software Ltd., www.imperas.com
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
 * either express or implied.
 *
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

// VMI header files
#include "vmi/vmiCxt.h"
#include "vmi/vmiMessage.h"
#include "vmi/vmiMt.h"
#include "vmi/vmiRt.h"

// model header files
#include "riscvBlockState.h"
#include "riscvDecode.h"
#include "riscvExceptions.h"
#include "riscvFunctions.h"
#include "riscvMessage.h"
#include "riscvMode.h"
#include "riscvStructure.h"
#include "riscvUtils.h"
#include "riscvVariant.h"
#include "riscvVM.h"


//
// Update the currently-enabled architecture settings
//
void riscvSetCurrentArch(riscvP riscv) {

    Uns32 MXL = RD_CSR_FIELD(riscv, misa, MXL);
    Bool  FS  = (RD_CSR_FIELD(riscv, mstatus, FS) != 0);

    // derive new architecture value based on misa value, preserving rounding
    // mode invalid setting
    riscvArchitecture arch = (
        (riscv->currentArch & ISA_RM_INVALID) |
        RD_CSR_FIELD(riscv, misa, Extensions) |
        (MXL<< XLEN_SHIFT)                    |
        (FS << MSTATUS_FS_SHIFT)
    );

    // mstatus.FS=0 disables floating point extensions
    if(!FS) {
        arch &= ~ISA_DF;
    }

    // mstatus.VS=0 disables vector extensions (if implemented)
    if(arch & ISA_V) {

        Uns32 WM_mstatus_VS = 0;

        // get mask of dirty bits for mstatus.VS in either 0.8 or 0.9 location
        if(riscvVFSupport(riscv, RVVF_VS_STATUS_8)) {
            WM_mstatus_VS = WM_mstatus_VS_8;
        } else if(riscvVFSupport(riscv, RVVF_VS_STATUS_9)) {
            WM_mstatus_VS = WM_mstatus_VS_9;
        }

        if(WM_mstatus_VS && !(RD_CSR(riscv, mstatus) & WM_mstatus_VS)) {
            arch &= ~ISA_V;
        }
    }

    // handle big endian access if required
    if(riscv->checkEndian && riscvGetCurrentDataEndian(riscv)) {
        arch |= ISA_BE;
    }

    if(riscv->currentArch != arch) {

        vmiProcessorP processor = (vmiProcessorP)riscv;

        // update current block mask to match architecture
        vmirtSetBlockMask(processor, arch);

        // update current architecture on processor
        riscv->currentArch = arch;
    }
}

//
// Return the configured XLEN (may not be the current XLEN if dynamic update
// of XLEN is allowed)
//
Uns32 riscvGetXlenArch(riscvP riscv) {

    riscvArchitecture arch   = riscv->configInfo.arch;
    Uns32             result = 0;

    if(arch & ISA_XLEN_64) {
        result = 64;
    } else if(arch & ISA_XLEN_32) {
        result = 32;
    } else {
        VMI_ABORT("invalid XLEN"); // LCOV_EXCL_LINE
    }

    return result;
}

//
// Return the configured FLEN (may not be the current FLEN if dynamic update
// of FLEN is allowed)
//
Uns32 riscvGetFlenArch(riscvP riscv) {

    riscvArchitecture arch   = riscv->configInfo.arch;
    Uns32             result = 0;

    if(arch & ISA_D) {
        result = 64;
    } else if(arch & ISA_F) {
        result = 32;
    }

    return result;
}

//
// Register extension callback block for the id with the base model
//
void riscvRegisterExtCB(riscvP riscv, riscvExtCBP extCB, Uns32 id) {

    riscvExtCBPP tail = &riscv->extCBs;
    riscvExtCBP  this;

    while((this=*tail)) {
        tail = &this->next;
    }

    *tail = extCB;
    extCB->next = 0;
    extCB->id   = id;
}

//
// Return the indexed extension's extCB clientData
//
void *riscvGetExtClientData(riscvP riscv, Uns32 id) {

    riscvExtCBP this = riscv->extCBs;

    while(this && (this->id!=id)) {
        this = this->next;
    }

    return this ? this->clientData : 0;
}

//
// Return extension configuration with the given id
//
riscvExtConfigCP riscvGetExtConfig(riscvP riscv, Uns32 id) {

    riscvExtConfigCPP extCfgs = riscv->configInfo.extensionConfigs;
    riscvExtConfigCP  extCfg  = 0;

    if(extCfgs) {
        while((extCfg=*extCfgs) && (extCfg->id!=id)) {
            extCfgs++;
        }
    }

    return extCfg;
}

//
// Return the current XLEN
//
Uns32 riscvGetXlenMode(riscvP riscv) {

    riscvArchitecture arch   = riscv->currentArch;
    Uns32             result = 0;

    if(arch & ISA_XLEN_64) {
        result = 64;
    } else if(arch & ISA_XLEN_32) {
        result = 32;
    } else {
        VMI_ABORT("invalid XLEN"); // LCOV_EXCL_LINE
    }

    return result;
}

//
// Does the processor support configurable endianness?
//
Bool riscvSupportEndian(riscvP riscv) {
    return (RISCV_PRIV_VERSION(riscv)>RVPV_20190405);
}

//
// Return endianness for data access in the given mode
//
memEndian riscvGetDataEndian(riscvP riscv, riscvMode mode) {

    memEndian result = riscv->dendian;

    if(!riscvSupportEndian(riscv)) {
        // no action
    } else if(mode==RISCV_MODE_USER) {
        result = RD_CSR_FIELD(riscv, mstatus, UBE);
    } else if(mode==RISCV_MODE_SUPERVISOR) {
        result = RD_CSR_FIELD_ALT(riscv, mstatush, mstatus, SBE);
    } else if(mode==RISCV_MODE_MACHINE) {
        result = RD_CSR_FIELD_ALT(riscv, mstatush, mstatus, MBE);
    } else {
        VMI_ABORT("invalid mode"); // LCOV_EXCL_LINE
    }

    return result;
}

//
// Return endianness for data access in the current mode
//
memEndian riscvGetCurrentDataEndian(riscvP riscv) {
    return riscvGetDataEndian(riscv, riscv->dmode);
}

//
// Return endianness of access
//
VMI_ENDIAN_FN(riscvGetEndian) {

    riscvP riscv = (riscvP)processor;

    if(isFetch) {
        return riscv->iendian;
    } else {
        return riscvGetCurrentDataEndian(riscv);
    }
}

//
// Return next PC after the given PC
//
VMI_NEXT_PC_FN(riscvNextPC) {

    // calculate nextPC ignoring wrapping
    riscvP riscv  = (riscvP)processor;
    Uns64  nextPC = thisPC + riscvGetInstructionSize(riscv, thisPC);

    // allow for wrapping if XLEN is 32
    if(riscvGetXlenMode(riscv)==32) {
        nextPC &= 0xffffffff;
    }

    return nextPC;
}

//
// Table of processor mode descriptions
//
// NOTE: The CPU Helper depends on the low order 2 bits of the code
//       matching the privilege level encoding for the mstatus.MPP field
//       defined in table 1.1 of RISC-V Privileged Architectures V1.10
//
static const vmiModeInfo modes[] = {

    [RISCV_MODE_USER] = {
        .name        = "User",
        .code        = RISCV_MODE_USER,
        .description = "User mode"
    },

    [RISCV_MODE_SUPERVISOR] = {
        .name        = "Supervisor",
        .code        = RISCV_MODE_SUPERVISOR,
        .description = "Supervisor mode"
    },

    [RISCV_MODE_HYPERVISOR] = {
        .name        = "Hypervisor",
        .code        = RISCV_MODE_HYPERVISOR,
        .description = "Hypervisor mode"
    },

    [RISCV_MODE_MACHINE] = {
        .name        = "Machine",
        .code        = RISCV_MODE_MACHINE,
        .description = "Machine mode"
    },

    [RISCV_MODE_DEBUG] = {
        .name        = "Debug",
        .code        = RISCV_MODE_DEBUG,
        .description = "Debug mode"
    },

    // terminator
    {0}
};

//
// Get mode name for the indexed mode
//
const char *riscvGetModeName(riscvMode mode) {
    return modes[mode].name;
}

//
// Get current mode
//
VMI_GET_MODE_FN(riscvGetMode) {

    riscvP riscv = (riscvP)processor;

    return &modes[inDebugMode(riscv) ? RISCV_MODE_DEBUG : getCurrentMode(riscv)];
}

//
// Change processor mode
//
void riscvSetMode(riscvP riscv, riscvMode mode) {

    riscvDMode dMode = mode;

    // if executing in supervisor or user mode, include VM-enabled indication
    if((mode<=RISCV_MODE_SUPERVISOR) && (RD_CSR_FIELD(riscv, satp, MODE))) {
        dMode |= RISCV_DMODE_VM;
    } else {
        dMode &= ~RISCV_DMODE_VM;
    }

    // update mode if it has changed
    if(riscv->mode != dMode) {
        riscv->mode = dMode;
        vmirtSetMode((vmiProcessorP)riscv, dMode);
    }

    // refresh current data domain (may be modified by mstatus.MPRV, and may
    // have changed while taking an exception even if mode has not changed)
    riscvVMRefreshMPRVDomain(riscv);

    // set step breakpoint if required
    riscvSetStepBreakpoint(riscv);
}

//
// Return the minimum supported processor mode
//
riscvMode riscvGetMinMode(riscvP riscv) {

    if(riscv->configInfo.arch & ISA_U) {
        return RISCV_MODE_USER;
    } else if(riscv->configInfo.arch & ISA_S) {
        return RISCV_MODE_SUPERVISOR;
    } else {
        return RISCV_MODE_MACHINE;
    }
}

//
// Does the processor implement the given mode?
//
Bool riscvHasMode(riscvP riscv, riscvMode mode) {

    switch(mode) {
        case RISCV_MODE_USER:
            return riscv->configInfo.arch & ISA_U;
        case RISCV_MODE_SUPERVISOR:
            return riscv->configInfo.arch & ISA_S;
        case RISCV_MODE_MACHINE:
            return True;
        case RISCV_MODE_DEBUG:
            return riscv->configInfo.debug_mode;
        default:
            return False;
    }
}

//
// Iterate processor modes
//
VMI_MODE_INFO_FN(riscvModeInfo) {

    riscvP riscv = (riscvP)processor;

    // on the first call, start with the first member of the table
    if(!prev) {
        prev = modes-1;
    }

    // get the next member
    vmiModeInfoCP this = prev+1;

    // skip to the next implemented mode
    while((this->name) && !riscvHasMode(riscv, this->code)) {
        this++;
    }

    // return the next member, or NULL if at the end of the list
    return (this->name) ? this : 0;
}

//
// Return the indexed X register name
//
const char *riscvGetXRegName(Uns32 index) {

    static const char *map[32] = {
        [0] = "zero",
        [1] = "ra",
        [2] = "sp",
        [3] = "gp",
        [4] = "tp",
        [5] = "t0",
        [6] = "t1",
        [7] = "t2",
        [8] = "s0",
        [9] = "s1",
        [10] = "a0",
        [11] = "a1",
        [12] = "a2",
        [13] = "a3",
        [14] = "a4",
        [15] = "a5",
        [16] = "a6",
        [17] = "a7",
        [18] = "s2",
        [19] = "s3",
        [20] = "s4",
        [21] = "s5",
        [22] = "s6",
        [23] = "s7",
        [24] = "s8",
        [25] = "s9",
        [26] = "s10",
        [27] = "s11",
        [28] = "t3",
        [29] = "t4",
        [30] = "t5",
        [31] = "t6",
    };

    // sanity check index is in range
    VMI_ASSERT(index<32, "Illegal index %u", index);

    return map[index];
}

//
// Return the indexed F register name
//
const char *riscvGetFRegName(Uns32 index) {

    static const char *map[32] = {
        [0] = "ft0",
        [1] = "ft1",
        [2] = "ft2",
        [3] = "ft3",
        [4] = "ft4",
        [5] = "ft5",
        [6] = "ft6",
        [7] = "ft7",
        [8] = "fs0",
        [9] = "fs1",
        [10] = "fa0",
        [11] = "fa1",
        [12] = "fa2",
        [13] = "fa3",
        [14] = "fa4",
        [15] = "fa5",
        [16] = "fa6",
        [17] = "fa7",
        [18] = "fs2",
        [19] = "fs3",
        [20] = "fs4",
        [21] = "fs5",
        [22] = "fs6",
        [23] = "fs7",
        [24] = "fs8",
        [25] = "fs9",
        [26] = "fs10",
        [27] = "fs11",
        [28] = "ft8",
        [29] = "ft9",
        [30] = "ft10",
        [31] = "ft11",
    };

    // sanity check index is in range
    VMI_ASSERT(index<32, "Illegal index %u", index);

    return map[index];
}

//
// Return the indexed V register name
//
const char *riscvGetVRegName(Uns32 index) {

    static const char *map[32] = {
        [0] = "v0",
        [1] = "v1",
        [2] = "v2",
        [3] = "v3",
        [4] = "v4",
        [5] = "v5",
        [6] = "v6",
        [7] = "v7",
        [8] = "v8",
        [9] = "v9",
        [10] = "v10",
        [11] = "v11",
        [12] = "v12",
        [13] = "v13",
        [14] = "v14",
        [15] = "v15",
        [16] = "v16",
        [17] = "v17",
        [18] = "v18",
        [19] = "v19",
        [20] = "v20",
        [21] = "v21",
        [22] = "v22",
        [23] = "v23",
        [24] = "v24",
        [25] = "v25",
        [26] = "v26",
        [27] = "v27",
        [28] = "v28",
        [29] = "v29",
        [30] = "v30",
        [31] = "v31",
    };

    // sanity check index is in range
    VMI_ASSERT(index<32, "Illegal index %u", index);

    return map[index];
}

//
// Utility function returning a vmiReg object to access the indexed vector
// register
//
vmiReg riscvGetVReg(riscvP riscv, Uns32 index) {

    void *value = &riscv->v[index*riscv->configInfo.VLEN/32];

    return vmimtGetExtReg((vmiProcessorP)riscv, value);
}

//
// Return index for the first feature identified by the given feature id
//
static Uns32 getFeatureIndex(riscvArchitecture feature) {

    Uns32 index = -1;

    // select first indicated feature
    feature = feature & -feature;

    // convert to index
    while(feature) {
        feature >>= 1;
        index++;
    }

    return index;
}

//
// Get character identifier for the first feature identified by the given
// feature id
//
char riscvGetFeatureChar(riscvArchitecture feature) {

    return getFeatureIndex(feature)+'A';
}

//
// Get description for the first feature identified by the given feature id
//
const char *riscvGetFeatureName(riscvArchitecture feature) {

    // table mapping to feature descriptions
    static const char *featureDescs[32] = {
        [RISCV_FEATURE_INDEX(XLEN32_CHAR)] = "32-bit XLEN",
        [RISCV_FEATURE_INDEX(XLEN64_CHAR)] = "64-bit XLEN",
        [RISCV_FEATURE_INDEX('A')]         = "extension A (atomic instructions)",
        [RISCV_FEATURE_INDEX('C')]         = "extension C (compressed instructions)",
        [RISCV_FEATURE_INDEX('E')]         = "RV32E base ISA",
        [RISCV_FEATURE_INDEX('D')]         = "extension D (double-precision floating point)",
        [RISCV_FEATURE_INDEX('F')]         = "extension F (single-precision floating point)",
        [RISCV_FEATURE_INDEX('I')]         = "RV32I/64I/128I base ISA",
        [RISCV_FEATURE_INDEX('M')]         = "extension M (integer multiply/divide instructions)",
        [RISCV_FEATURE_INDEX('N')]         = "extension N (user-level interrupts)",
        [RISCV_FEATURE_INDEX('S')]         = "extension S (Supervisor mode)",
        [RISCV_FEATURE_INDEX('U')]         = "extension U (User mode)",
        [RISCV_FEATURE_INDEX('V')]         = "extension V (vector instructions)",
        [RISCV_FEATURE_INDEX('X')]         = "extension X (non-standard extensions present)"
    };

    // get feature description
    return featureDescs[getFeatureIndex(feature)];
}

//
// Parse the extensions string
//
riscvArchitecture riscvParseExtensions(const char *extensions) {

    riscvArchitecture result = 0;

    if(extensions) {

        const char *tail = extensions;
        Bool        ok   = True;
        char        extension;

        while(ok && (extension=*tail++)) {

            ok = (extension>='A') && (extension<='Z');

            if(!ok) {
                vmiMessage("E", CPU_PREFIX"_ILLEXT",
                    "Illegal extension string \"%s\" - letters A-Z required",
                    extensions
                );
            } else {
                result |= (1<<(extension-'A'));
            }
        }
    }

    return result;
}


////////////////////////////////////////////////////////////////////////////////
// PROCESSOR RUN STATE TRANSITION HANDLING
////////////////////////////////////////////////////////////////////////////////

//
// If this memory access callback is triggered, abort any active load linked
//
static VMI_MEM_WATCH_FN(abortEA) {
    if(processor) {
        riscvAbortExclusiveAccess((riscvP)userData);
    }
}

//
// Install or remove the exclusive access monitor callback
//
static void updateExclusiveAccessCallback(riscvP riscv, Bool install) {

    memDomainP domain  = vmirtGetProcessorDataDomain((vmiProcessorP)riscv);
    Uns64      simLow  = riscv->exclusiveTag;
    Uns64      simHigh = simLow + ~riscv->exclusiveTagMask;

    // install or remove a watchpoint on the current exclusive access address
    if(install) {
        vmirtAddWriteCallback(domain, 0, simLow, simHigh, abortEA, riscv);
    } else {
        vmirtRemoveWriteCallback(domain, 0, simLow, simHigh, abortEA, riscv);
    }
}

//
// Abort any active exclusive access
//
void riscvAbortExclusiveAccess(riscvP riscv) {

    if(riscv->exclusiveTag != RISCV_NO_TAG) {

        // remove callback on exclusive access monitor region
        updateExclusiveAccessCallback(riscv, False);

        // clear exclusive tag (AFTER updateExclusiveAccessCallback)
        riscv->exclusiveTag = RISCV_NO_TAG;
    }
}

//
// Install or remove the exclusive access monitor callback if required
//
void riscvUpdateExclusiveAccessCallback(riscvP riscv, Bool install) {
    if(riscv->exclusiveTag != RISCV_NO_TAG) {
        updateExclusiveAccessCallback(riscv, install);
    }
}

//
// This is called on simulator context switch (when this processor is either
// about to start or about to stop simulation)
//
VMI_IASSWITCH_FN(riscvContextSwitchCB) {

    riscvP      riscv = (riscvP)processor;
    riscvExtCBP extCB;

    riscvUpdateExclusiveAccessCallback(riscv, state==RS_SUSPEND);

    // call derived model context switch function if required
    for(extCB=riscv->extCBs; extCB; extCB=extCB->next) {
        if(extCB->switchCB) {
            extCB->switchCB(riscv, state, extCB->clientData);
        }
    }
}


////////////////////////////////////////////////////////////////////////////////
// TRANSACTION MODE
////////////////////////////////////////////////////////////////////////////////

//
// Enable or disable transaction mode
//
RISCV_SET_TMODE_FN(riscvSetTMode) {

    // flush dictionaries the first time transaction mode is enabled
    if(enable && !riscv->useTMode) {

        riscv->useTMode = True;

        vmirtFlushAllDicts((vmiProcessorP)riscv);
    }

    // enable mode using polymorphic key
    if(enable) {
        riscv->pmKey |= PMK_TRANSACTION;
    } else {
        riscv->pmKey &= ~PMK_TRANSACTION;
    }
}

//
// Return true if in transaction mode
//
RISCV_GET_TMODE_FN(riscvGetTMode) {
    return (riscv->pmKey & PMK_TRANSACTION) != 0;
}

